reframe/core/hooks.py

```diff
@@ -170,10 +170,15 @@ def __contains__(self, key):
     def __getattr__(self, name):
         return getattr(self.__hooks, name)
 
-    def update(self, hooks):
+    def update(self, hooks, *, denied_hooks={}):
+        '''Update the hook registry with the hooks from another hook registry.
+
+        The optional ``denied_hooks`` argument takes a set of disallowed
+        hook names, preventing their inclusion into the current hook registry.
+        '''
         for phase, hks in hooks.items():
             self.__hooks.setdefault(phase, util.OrderedSet())
-            for h in hks:
+            for h in (hk for hk in hks if hk.__name__ not in denied_hooks):
                 self.__hooks[phase].add(h)
```

Comment:
```suggestion
            for h in hks:
                if h.__name__ not in denied_hooks:
                    self.__hooks[phase].add(h)
```

We only need to iterate over `hks` once.